### DISCUSSION TITLE
WGSL compile time constants & extensibility

### DISCUSSION THREAD

**Discussion by AshleyScirra** [gpuweb/gpuweb#1392] [Ideas/Feedback]
I've been gradually porting our 2D web game engine [Construct](https://www.construct.net) from WebGL to WebGPU. One potential issue I foresee is how WGSL handles extensibility. Our engine - like some others - allows third-party developers to write their own shaders for custom effects. These shaders need to combine both engine built-ins and their own custom features.

In GLSL we have a preprocessor which largely solves this, and also with WebGL 1 style uniforms things tend to be pretty relaxed anyway - e.g. you can just paste in a uniform like `uniform float engine_param;`, and if it's there it works. WGSL is stricter in that you have to define things like the precise layout of uniform buffers.

What we want to avoid is hard-coding things in WGSL that then get distributed by third-party developers. For example consider the following WGSL fragment shader that a third-party developer might customise:

```
[[block]] struct Uniforms {
	[[offset(0})]] engine_parameter: f32;

	[[offset(128)]] custom_parameter: f32;
};
[[binding(1), set(0)]] var<uniform> uniforms : Uniforms;

[[location(0)]] var<out> outColor : vec4<f32>;

[[stage(fragment)]]
fn main() -> void {
	outColor = myFancyEffect(uniforms.engine_parameter, uniforms.custom_parameter);
	return;
}`;
```

The third-party developer writes `myFancyEffect`, which combines both an engine-provided value (`engine_parameter`), and a custom value specific to their effect (`custom_parameter`). Suppose they then distribute this effect. Then because all sorts of things are hard-coded, changing any of the following in our engine becomes a breaking change:

- the offset of `engine_parameter`
- the type of `engine_parameter` (e.g. what if we wanted to conditionally make it `f16` in future?)
- the offset that custom parameters start from
- the binding and set of `uniforms`
- the location of `outColor`

This imposes some serious constraints on our ability to ever change things like the pipeline layout, the uniform block layout etc. and additionally makes it harder to use f16 types when those are supported. If we have compile-time constants though, we can distribute something like this:

```
%%ENGINE_BUILTINS%%

[[block]] struct Uniforms {
	[[offset(OFFSET_ENGINE_PARAMETER)]] engine_parameter: TYPE_ENGINE_PARAM;

	[[offset(OFFSET_CUSTOM_PARAMETER)]] custom_parameter: f32;
};
[[binding(BINDING_UNIFORMS), set(SET_UNIFORMS)]] var<uniform> uniforms : Uniforms;

[[location(LOCATION_OUTCOLOR)]] var<out> outColor : TYPE_OUTCOLOR;

[[stage(fragment)]]
fn main() -> void {
	outColor = myFancyEffect(uniforms.engine_parameter, uniforms.custom_parameter);
	return;
}`;
```

Then using preprocessor syntax to demonstrate what our engine inserts, `%%ENGINE_BUILTINS%%` could be substituted with something like:

```
#define OFFSET_ENGINE_PARAMETER 0
#define OFFSET_CUSTOM_PARAMETER 128
#define TYPE_ENGINE_PARAM f32
#define TYPE_OUTCOLOR vec4<f32>
#define BINDING_UNIFORMS 1
#define SET_UNIFORMS 0
#define LOCATION_OUTCOLOR 0
```

Since our engine controls these parameters on startup, we can then generate the right values for our engine. This then affords flexibility to do things like rearrange the uniform block layout, change the pipeline layout, change the types of engine built-ins, alter the output location, and so on. We could make these changes after the distribution of third-party shaders without breaking them. A related aspect is also that changing f32 to f16 changes the size of uniforms and so could have knock-on effects for the offsets of uniforms after it - so changing type also changes offsets.

The question is what we will be able to do with this in the MVP. If we can't do this in the MVP, we are forced in to hard-coding and the breaking changes situation. So I'd argue we need to support this one way or another in the MVP.

Some related things appear to be:
- type aliases - ideally we could define something like `f32_or_16t` (hopefully with a better name...) allowing engine builtins to use f16 in future if supported, else fall back to f32
- compile-time constants for uniform offsets - related to issue #1272 - although there appears to be discussion about removing offsets which would make this harder (#1349, #1303, #1361)
- compile-time constants also in binding, set, and location (and others?)

So in short my takeaways would be:
- please keep offset in some form
- ship compile-time constants in MVP with support for offset, binding, set, location etc.
- ship type aliases in MVP

**Comment by dj2** [comment]
There is current ongoing discussion about struct layouts. The current thinking is to remove `offset` and replace it with ... something else. That something has not been defined yet. Using `offset` while affording a lot of flexibility, has a large downside of adding a param requires renumbering everything after it. This makes working with the structs a lot more difficult then they need to be.

The values used in `offset`, `binding`, `group`, etc are all integers at this point. We also, purposely, don't have a preprocessor. That said, we do have Javascript. The thinking being, if you want to do preprocessor type work, or compute things ahead of time, it can be done in Javascript and the value set directly into the shader. This would also work for your f32 -> f16 flip, potentially, by defining the type in JS and substituting into the shader.

Type aliases are already in the language and I don't see any desire to remove them, so they should be there for MVP.

**Reply by AshleyScirra** [reply]
The difference here is with third-party shader distribution, you cannot use JavaScript to write values in to the shader. The WGSL code is coming from an outside source, it's not a template string in our own code that we control. Therefore we cannot substitute things inside it. Unless we effectively write our own preprocessor, which I really don't want to have to do.

**Reply by dj2** [reply]
Ah, I see, yea that sounds like it would need some kind of preprocessor to handle. @Kangz has previously mentioned it shouldn't be too hard to create a JS preprocessor which could be run before handing the WGSL shader off to WebGPU.

**Comment by kvark** [comment]
This is a very detailed feedback, thank you for writing it down!
For the most part, your problem comes from the need to fill out the `offset` decorations. It's a pain, but not one that the group has fully acknowledged yet. There are proposals like #1361 and #1393, which remove it from the API surface, and make it internal (similar  to HLSL/MSL/GLSL).

In addition, you may consider a more rigid approach for extensibility. For example, you can say that bind group 0 is for the engine, bind group 1 is for the user bindings, and bind group 2 is for the per-draw bindings. It's less powerful, but isolation at the bind group level (as well as fixed bind group IDs) allows the code that engine doesn't know about to be fully self-contained and independent.

Compile-time constants for locations and bindings are something we could consider (post-MVP) and even implement (with some caveats on SPIR-V), but it would look a bit strange. Generally, I'd expect the decorations `[[xxx(yy)]]` to be on level above the shader code, they are hook it up to the pipeline state. But if we allow writing `[[location(my_constant)]]` then the relationship between the decorations and non-decorations becomes more complex.

**Reply by AshleyScirra** [reply]
> For the most part, your problem comes from the need to fill out the `offset` decorations... There are proposals like #1361 and #1393, which remove it from the API surface

The thing is we actually need `offset`, or some equivalent, to make sure we can do things like offset third-party parameters from engine parameters. In other words we want something like "third-party parameters begin from offset 128" without having to define what occupies the first 128 bytes.

> In addition, you may consider a more rigid approach for extensibility. For example, you can say that bind group 0 is for the engine, bind group 1 is for the user bindings, and bind group 2 is for the per-draw bindings.

This could work, but since the spec only guarantees up to 4 bind groups, it seems a shame to have to waste one of such a limited resource over this. Plus it means we still have to hard-code a group and then we can never change it without breaking things, which is what I was hoping to avoid in the first place.

**Reply by kvark** [reply]
Using an `offset` for this is an interesting take. But did you consider just making this a separate uniform buffer binding instead?

**Reply by AshleyScirra** [reply]
I guess we could. Isn't it more efficient to update everything in one buffer, especially if it's a relatively small amount of information? We also still have the problem of hard-coding which binding that would be.

**Reply by kvark** [reply]
The separate bindings do not require you to have separate buffers. It can be the same buffer, just different sub-ranges bound to different bindings. So there is no efficiency difference here.

**Reply by dneto0** [reply]
First, thank you for the detailed and thoughtful feedback!

> The thing is we actually need offset, or some equivalent, to make sure we can do things like offset third-party parameters from engine parameters. In other words we want something like "third-party parameters begin from offset 128" without having to define what occupies the first 128 bytes.

I'm writing much later than the discussion above.  Offset has been removed, and in its place are default sizes and alignments, and optional explicit size and align attributes.

So one way to do address that use case is to apply a size attribute to the one you want to be 128 bytes big:

```
[[block]] struct AllParams {
   [[size(128)]] engineParams : EngineParamsStruct;
   thirdPartyParams : thirdPartyParamsStruct; // will start at offset 128. 
};
```
This would error out if EngineParamsStruct is larger than 128 bytes.
